cdap-common/src/main/java/co/cask/cdap/common/security/AbstractCachedUGIProvider.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.common.security;
+
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import com.google.common.base.Throwables;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import org.apache.hadoop.security.UserGroupInformation;
+
+import java.io.IOException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * An abstract base class for {@link UGIProvider} that provides caching of {@link UserGroupInformation}.
+ */
+public abstract class AbstractCachedUGIProvider implements UGIProvider {
+
+  private final LoadingCache<ImpersonationInfo, UserGroupInformation> ugiCache;
+
+  protected AbstractCachedUGIProvider(CConfiguration cConf) {
+    this.ugiCache = createUGICache(cConf);
+  }
+
+  /**
+   * Creates a new {@link UserGroupInformation} based on the given {@link ImpersonationInfo}.
+   */
+  protected abstract UserGroupInformation createUGI(ImpersonationInfo impersonationInfo) throws IOException;
+
+  @Override
+  public final UserGroupInformation getConfiguredUGI(ImpersonationInfo impersonationInfo) throws IOException {
+    try {
+      return ugiCache.get(impersonationInfo);
+    } catch (ExecutionException e) {
+      // Get the root cause of the failure
+      Throwable cause = Throwables.getRootCause(e);
+      // Propagate iff the cause is an IOException
+      Throwables.propagateIfInstanceOf(cause, IOException.class);
+      // Otherwise always wrap it with IOException
```

Comment:
I changed the underlying implementation not to throw RuntimeException explicitly (e.g. wrapping malformed URL exception, which is an IOException). Also changed the propagation to propagate runtime exception without wrapping. This at least keep runtime exception truly runtime exception (meaning unexpected).